codec-http/src/main/java/io/netty/handler/codec/http/cache/CacheKeyGenerator.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2018 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http.cache;
+
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.util.internal.StringUtil;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+public class CacheKeyGenerator {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(CacheKeyGenerator.class);
+
+    public String generateKey(HttpRequest request) {
```

Comment:
HTTP cache keys are actually way more complicated than that.

The only correct cache key is the **full entity**: url + headers + body. I ran into lots of troubles naively only using the url:
* headers trigger completely different server behavior: `cookie`, `accept`, `accept-encoding`, `accept-language`, etc
* it's not clear in the spec if request body is legit for GET or not, but some products such as ElasticSearch have been using body to pass structured and possibly large request parameters, even for GET.

IMHO, you should make CacheKeyGenerator an interface and let the user provides its own implementation.

Reply:
Done